src/ext/default.c

```diff
@@ -83,40 +104,53 @@ static double TFIDFNormDocLenScorer(const ScoringFunctionArgs *ctx, const RSInde
 
 /* recursively calculate score for each token, summing up sub tokens */
 static double bm25Recursive(const ScoringFunctionArgs *ctx, const RSIndexResult *r,
-                            const RSDocumentMetadata *dmd) {
+                            const RSDocumentMetadata *dmd, bool explain) {
   static const float b = 0.5;
   static const float k1 = 1.2;
   double f = (double)r->freq;
-
+  double ret = 0;
+  
   if (r->type == RSResultType_Term) {
     double idf = (r->term.term ? r->term.term->idf : 0);
 
-    double ret = idf * f / (f + k1 * (1.0f - b + b * ctx->indexStats.avgDocLen));
-    return ret;
-  }
-
-  if (r->type & (RSResultType_Intersection | RSResultType_Union)) {
-    double ret = 0;
+    ret = idf * f / (f + k1 * (1.0f - b + b * ctx->indexStats.avgDocLen));
+    EXPLAIN(r, "%.2f = IDF %.2f * F %d / (F %d + k1 1.2 * (1 - b 0.5 + b 0.5 * Average Len %.2f))", 
+            ret, idf, r->freq, r->freq, ctx->indexStats.avgDocLen);  
+  } else if (r->type & (RSResultType_Intersection | RSResultType_Union)) {
     for (int i = 0; i < r->agg.numChildren; i++) {
-      ret += bm25Recursive(ctx, r->agg.children[i], dmd);
+      ret += bm25Recursive(ctx, r->agg.children[i], dmd, explain);
     }
-    return r->weight * ret;
+    EXPLAIN(r, "%.2f = Weight %.2f * children BM25 %.2f",  r->weight * ret, r->weight, ret);
+    ret *= r->weight;
+  } else if (f) {   // default for virtual type -just disregard the idf
+    ret = r->weight * f / (f + k1 * (1.0f - b + b * ctx->indexStats.avgDocLen));
+    EXPLAIN(r, "%.2f = Weight %.2f * F %d / (F %d + k1 1.2 * (1 - b 0.5 + b 0.5 * Average Len %.2f))", 
+            ret, r->weight, r->freq, r->freq, ctx->indexStats.avgDocLen);
+  } else {
+    EXPLAIN(r, "Frequency 0 -> value 0");
   }
-  // default for virtual type -just disregard the idf
-  return r->weight * (r->freq ? f / (f + k1 * (1.0f - b + b * ctx->indexStats.avgDocLen)) : 0);
+
+  return ret;
 }
 
 /* BM25 scoring function */
 static double BM25Scorer(const ScoringFunctionArgs *ctx, const RSIndexResult *r,
                          const RSDocumentMetadata *dmd, double minScore) {
-  double score = dmd->score * bm25Recursive(ctx, r, dmd);
+  bool explain = ctx->scoreflags & SCORE_F_WITH_EXPLANATION;
```

Comment:
I think its better to just put explain variable inside the ctx and only update it once, what do you think?

Reply:
Yeah, I'd agree with that. Make the `ctx` contain an optional `char *` pointer. If the pointer is non-NULL on input, it will be updated with the score.

Reply:
After discussion, we decided to leave as an option for multiple flags in the future.